src/sentry/tagstore/snuba/backend.py

```diff
@@ -172,18 +176,34 @@ def __get_tag_keys_for_projects(
             aggregations.append(["uniq", "tags_value", "values_seen"])
         conditions = []
 
-        result = snuba.query(
-            start=start,
-            end=end,
-            groupby=["tags_key"],
-            conditions=conditions,
-            filter_keys=filters,
-            aggregations=aggregations,
-            limit=limit,
-            orderby="-count",
-            referrer="tagstore.__get_tag_keys",
-            **kwargs
-        )
+        filtering_strings = [u"{}={}".format(key, value) for key, value in six.iteritems(filters)]
+
+        cache_key = u"tagstore.__get_tag_keys:{}".format(md5_text(*filtering_strings).hexdigest())
+        # Round times by 5 minutes since we cache for 5 minutes
+        if start:
+            start = datetime.datetime(
+                start.year, start.month, start.day, start.hour, start.minute // 5 * 5
+            )
+            cache_key += ":" + start.isoformat()
+        if end:
+            end = datetime.datetime(end.year, end.month, end.day, end.hour, end.minute // 5 * 5)
+            cache_key += ":" + end.isoformat()
+
+        result = cache.get(cache_key, None)
+        if result is None:
+            result = snuba.query(
+                start=start,
+                end=end,
+                groupby=["tags_key"],
+                conditions=conditions,
+                filter_keys=filters,
+                aggregations=aggregations,
+                limit=limit,
+                orderby="-count",
+                referrer="tagstore.__get_tag_keys",
+                **kwargs
+            )
+            cache.set(cache_key, result, 300)
```

Comment:
Do you have an estimate on the amount of data you are going to cache? This may have an impact on production.

Reply:
I don't, this would be done by creating a separate change using `sentry.utils.metrics` first right?

Reply:
That's only part of it. We need these:
- the amount of cache unique keys
- the average cache value size.

For the first one I think the only way is to start caching for a very small percentage of the requests so we can estimate how many entries we will have.
For the second we need to bump a metric with the seize of the value we want to store so that we have the average.

Also the ops team needs to be in the loop before you start caching.

Reply:
> for a very small percentage of the requests
Is that by getting this into Canary? Or do I throw some RNG here to only cache sometimes.

Reply:
The second, unless you are able to estimate the size of the cache in some different way.

Reply:
Could this start off as just caching a boolean on whether or not we would have hit the cache, and also emitting a metric to let us see the cache miss rates + and response size. That would let you tweak the cache key to get a good hit rate before caching any actual data.

Reply:
Sure we should still have an idea of the amount of cache keys, so we need to start with a fraciton of the traffic that we can bump up very quickly if, as I would expect, this is not going to have a majore impact on memecache

Reply:
Hmm I like that solution @markstory, but I'm not sure at what level of caching will result in impacts on production. If I use your Boolean method will I still have to limit this to a small percentage? Or can I unleash it on everything?